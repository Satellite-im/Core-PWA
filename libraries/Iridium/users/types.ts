import { CID } from 'multiformats'

export type User = {
  did: string
  name: string
  status?: string
  peerId?: string
  seen?: number
  photoHash?: string | CID
  accountUrl?: string
  about?: string
  isWelcomeMessage?: boolean
}

export type UserStatus = 'online' | 'offline' | 'busy' | 'away' | 'mobile'

export const UsersError = {
  NETWORK_ERROR: 'errors.users.network',
  USER_NOT_FOUND: 'errors.users.not_found',
}
