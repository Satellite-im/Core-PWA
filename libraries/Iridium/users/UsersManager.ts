import Vue from 'vue'
import { v4 } from 'uuid'
import {
  IridiumPeerIdentifier,
  Emitter,
  didUtils,
  encoding,
  IridiumPubsubMessage,
} from '@satellite-im/iridium'
import type {
  IridiumMessage,
  IridiumGetOptions,
  IridiumSetOptions,
} from '@satellite-im/iridium/src/types'
import { IridiumDecodedPayload } from '@satellite-im/iridium/src/core/encoding'
import type { IridiumManager } from '../IridiumManager'
import { User, UsersError, UserStatus } from './types'
import logger from '~/plugins/local/logger'
import { FriendRequestStatus } from '~/libraries/Iridium/friends/types'

export type IridiumUserEvent = {
  to: IridiumPeerIdentifier
  status: FriendRequestStatus
  user: User
  data?: any
  at: number
}

export type UserState = { [key: User['did']]: User }

export type IridiumUserPubsub = IridiumPubsubMessage<
  IridiumDecodedPayload<IridiumUserEvent>
>

export default class UsersManager extends Emitter<IridiumUserPubsub> {
  public readonly iridium: IridiumManager
  public state: UserState = {}
  public userStatus: { [key: string]: UserStatus } = {}

  private loggerTag = 'iridium/users'

  constructor(iridium: IridiumManager) {
    super()
    this.iridium = iridium
  }

  get list() {
    return Object.values(this.state)
  }

  async init() {
    const iridium = this.iridium.connector
    if (!iridium) {
      throw new Error('cannot initialize users, no iridium connector')
    }

    logger.log(this.loggerTag, 'initializing')
    const pubsub = iridium.pubsub.subscriptions()
    logger.info(this.loggerTag, 'pubsub', pubsub)
    await this.fetch()
    logger.log(this.loggerTag, 'users state loaded', this.state)
    logger.info(this.loggerTag, 'subscribing to announce topic')
    await iridium.subscribe<IridiumUserPubsub>('/users/announce', {
      handler: this.onUsersAnnounce.bind(this),
      sync: true,
    })
    logger.log(this.loggerTag, 'listening for user activity', this.state)

    this.iridium.connector?.p2p.on('peer/disconnect', (peer) => {
      logger.info(this.loggerTag, 'peer disconnected', peer)
      this.setUser(peer.did, {
        ...this.state[peer.did],
        status: 'offline',
        seen: Date.now(),
      })
    })

    setInterval(() => {
      this.list.forEach(async (user: User) => {
        if (
          user.status === 'online' &&
          Number(user.seen) < Date.now() - 1000 * 30
        ) {
          logger.info(this.loggerTag, 'user timed out', user)
          this.iridium.users.setUser(user.did, {
            ...user,
            status: 'offline',
            seen: Date.now(),
          })
        }
      })
    }, 10000)

    this.iridium.connector?.p2p.on(
      'node/message/sync/searchPeer',
      (message) => {
        const peers = message.payload.body.peers as User[]
        peers.forEach((peer) => {
          this.setUser(peer.did, peer)
        })
        this.emit(
          `searchResults${
            message.payload.body.request
              ? `/${message.payload.body.request}`
              : ''
          }`,
          peers,
        )
      },
    )

    logger.info(this.loggerTag, 'initialized', this)
    this.emit('ready', {})
  }

  async unsubscribe() {
    await this.iridium.connector?.unsubscribe(`/users/announce`)
  }

  /**
   * @method fetch
   * @description fetch remote state
   * @returns updated state
   */
  async fetch() {
    const fetched = await this.get('/')
    this.state = Object.keys(fetched).reduce((acc, key: string) => {
      acc[key] = { ...fetched[key], status: 'offline' }
      return acc
    }, this.state)
  }

  async getUsers(): Promise<{ [key: string]: User }> {
    return this.state || {}
  }

  private async onUsersAnnounce(message: IridiumUserPubsub) {
    const { from, payload } = message
    const { status } = payload.body

    if (status === 'removed') {
      await this.userRemove(from)
    }
  }

  /**
   * @method get
   * @description get remote state
   * @param path string (required)
   * @param options object
   * @returns iridium's connector result
   */
  get<T = any>(path: string, options: IridiumGetOptions = {}): Promise<T> {
    if (!this.iridium.connector) {
      logger.error(this.loggerTag, 'network error')
      throw new Error(UsersError.NETWORK_ERROR)
    }
    return this.iridium.connector?.get<T>(
      `/users${path === '/' ? '' : path}`,
      options,
    )
  }

  /**
   * @method set
   * @description updates remote state
   * @param path string (required)
   * @param payload any (required)
   * @param options object
   * @returns iridium's connector result
   */
  set(path: string, payload: any, options: IridiumSetOptions = {}) {
    logger.info(this.loggerTag, 'path, paylaod and state', {
      path,
      payload,
      state: this.state,
    })

    return this.iridium.connector?.set(
      `/users${path === '/' ? '' : path}`,
      payload,
      options,
    )
  }

  /**
   * @method hasUser
   * @description check on the local state if there's a user with the current id
   * @param id string (required)
   * @returns boolean
   */
  hasUser(did: IridiumPeerIdentifier) {
    const id = didUtils.didString(did)
    logger.info(this.loggerTag, 'has user', {
      id,
      hasUser: id && !!this.state[id],
    })
    return id && !!this.state?.[id]
  }

  /**
   * @method getUser
   * @description check if there's a user with the current id and returns user data object
   * @param id string (required)
   * @returns user data object if found in the local state
   */
  getUser(did: IridiumPeerIdentifier): User | undefined {
    return this.state[didUtils.didString(did)]
  }

  async searchPeer(query: string, page: number = 0): Promise<User[]> {
    const iridium = this.iridium.connector
    if (!iridium) {
      return []
    }

    return new Promise<User[]>((resolve) => {
      const id = v4()
      if (!iridium.p2p.primaryNodeID) {
        return resolve([])
      }

      const timeout = setTimeout(() => {
        logger.warn(this.loggerTag, 'peer search timeout', { query })
        resolve([])
      }, 10000)

      this.once(`searchResults/${id}`, (results: User[]) => {
        clearTimeout(timeout)
        logger.info(this.loggerTag, 'peer search results', results)
        resolve(results)
      })

      logger.info(this.loggerTag, 'peer search query', { query, page, id })
      return iridium.p2p.send(iridium.p2p.primaryNodeID, {
        type: 'sync/searchPeer',
        id,
        query,
        page,
      })
    })
  }

  async setUser(id: IridiumPeerIdentifier, user: User) {
    const did = didUtils.didString(id)
    this.state[did] = user
    await this.set(`/${did}`, user)
  }

  setUserStatus(did: IridiumPeerIdentifier, status: UserStatus) {
    Vue.set(this.userStatus, didUtils.didString(did), status)
  }

  async send(event: IridiumUserEvent) {
    return this.iridium.connector?.publish(`/users/announce`, event, {
      encrypt: {
        recipients: [
          typeof event.to === 'string' ? event.to : event.to.id,
          this.iridium.connector?.id,
        ],
      },
    })
  }

  /**
   * @method userRemove
   * @description remove a user and announce it to the remote user
   * @param did - IridiumPeerIdentifier (required)
   * @returns Promise<void>
   */
  async userRemove(did: IridiumPeerIdentifier): Promise<void> {
    if (!this.iridium.connector) {
      logger.error(this.loggerTag, 'network error')
      throw new Error(UsersError.NETWORK_ERROR)
    }
    const profile = await this.iridium.profile?.get()
    if (!profile) {
      logger.error(this.loggerTag, 'network error')
      throw new Error(UsersError.NETWORK_ERROR)
    }

    const user = this.getUser(did)
    if (!user) {
      logger.error(this.loggerTag, 'user not found', { did })
      throw new Error(UsersError.USER_NOT_FOUND)
    }

    Vue.delete(this.state, didUtils.didString(did))
    await this.set(`/`, this.state)
    const id = await encoding.hash(
      [user.did, this.iridium.connector?.id].sort(),
    )
    if (id && this.iridium.chat.hasConversation(id)) {
      this.iridium.chat.deleteConversation(id)
    }

    logger.info(this.loggerTag, 'user removed', { did, user })

    // Announce to the remote user
    if (didUtils.didString(did) !== this.iridium.connector?.id) {
      const payload: IridiumUserEvent = {
        to: did,
        status: 'removed',
        at: Date.now(),
        user: {
          name: profile.name,
          did: profile.did,
        },
      }
      logger.info(this.loggerTag, 'announce remove user', {
        did,
        payload,
      })
      await this.send(payload)
    }
  }
}
