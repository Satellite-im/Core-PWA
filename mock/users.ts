export const Users = [
  {
    name: 'Taurus Nix',
    address: '0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f',
    status: 'Some important status message',
    state: 'online',
    last_message:
      'Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.',
  },
  {
    name: 'Halley Themis',
    address: '0x07ee55aa48bb72dcc6e9d78256648910de513eca',
    status: 'Working on the space station',
    state: 'offline',
    last_message:
      'At the edge of forever bits of moving fluff radio telescope the carbon in our apple pies are creatures of the cosmos rings of Uranus.',
  },
  {
    name: 'Lyra Cassini',
    address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
    status: 'Working on the space station',
    state: 'idle',
    last_message:
      'Tendrils of gossamer clouds stirred by starlight citizens of distant epochs network of wormholes corpus callosum the sky calls to us.',
  },
  {
    name: 'Phoenix Kalindi',
    address: '0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c',
    status: 'Working on the space station',
    state: 'online',
    last_message:
      'Ship of the imagination consciousness across the centuries network of wormholes finite but unbounded inconspicuous motes of rock and gas.',
    profile_picture:
      'https://st2.depositphotos.com/1009634/7235/v/600/depositphotos_72350117-stock-illustration-no-user-profile-picture-hand.jpg',
  },
  {
    name: 'Ariel Larissa',
    address: '0xdc76cd25977e0a5ae17155770273ad58648900d3',
    status: 'Working on the space station',
    state: 'idle',
    last_message:
      'Network of wormholes science citizens of distant epochs tingling of the spine cosmic ocean brain is the seed of intelligence. The sky calls to us two ghostly white figures in coveralls and helmets are softly dancing muse about paroxysm of global death the only home weve ever known hearts of the stars.',
    profile_picture:
      'https://st2.depositphotos.com/1009634/7235/v/600/depositphotos_72350117-stock-illustration-no-user-profile-picture-hand.jpg',
  },
]

export const User = {
  name: 'Lyra Cassini',
  address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
  status: 'Working on the space station',
  state: 'online',
  last_message:
    'Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.',
}

export const CallUsers = [
  {
    name: 'Taurus Nix',
    address: '0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f',
    status: 'Some important status message',
    state: 'online',
    last_message:
      'Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.',
  },
  {
    name: 'Halley Themis',
    address: '0x07ee55aa48bb72dcc6e9d78256648910de513eca',
    status: 'Working on the space station',
    state: 'offline',
    last_message:
      'At the edge of forever bits of moving fluff radio telescope the carbon in our apple pies are creatures of the cosmos rings of Uranus.',
  },
  // {
  //   name: 'Lyra Cassini',
  //   address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
  //   status: 'Working on the space station',
  //   state: 'idle',
  //   last_message:
  //     'Tendrils of gossamer clouds stirred by starlight citizens of distant epochs network of wormholes corpus callosum the sky calls to us.',
  // },
  // {
  //   name: 'Phoenix Kalindi',
  //   address: '0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c',
  //   status: 'Working on the space station',
  //   state: 'online',
  //   last_message:
  //     'Ship of the imagination consciousness across the centuries network of wormholes finite but unbounded inconspicuous motes of rock and gas.',
  //   profile_picture:
  //     'https://st2.depositphotos.com/1009634/7235/v/600/depositphotos_72350117-stock-illustration-no-user-profile-picture-hand.jpg',
  // },
  // {
  //   name: 'Ariel Larissa',
  //   address: '0xdc76cd25977e0a5ae17155770273ad58648900d3',
  //   status: 'Working on the space station',
  //   state: 'idle',
  //   last_message:
  //     'Network of wormholes science citizens of distant epochs tingling of the spine cosmic ocean brain is the seed of intelligence. The sky calls to us two ghostly white figures in coveralls and helmets are softly dancing muse about paroxysm of global death the only home weve ever known hearts of the stars.',
  //   profile_picture:
  //     'https://st2.depositphotos.com/1009634/7235/v/600/depositphotos_72350117-stock-illustration-no-user-profile-picture-hand.jpg',
  // },
  // {
  //   name: 'Lyra Cassini test',
  //   address: '0x9bf4001d307dfd62b26a2f1307ee0c0307632d59',
  //   status: 'Working on the space station',
  //   state: 'idle',
  //   last_message:
  //     'Tendrils of gossamer clouds stirred by starlight citizens of distant epochs network of wormholes corpus callosum the sky calls to us.',
  // },
  // {
  //   name: 'Phoenix Kalindi test',
  //   address: '0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c',
  //   status: 'Working on the space station',
  //   state: 'online',
  //   last_message:
  //     'Ship of the imagination consciousness across the centuries network of wormholes finite but unbounded inconspicuous motes of rock and gas.',
  //   profile_picture:
  //     'https://st2.depositphotos.com/1009634/7235/v/600/depositphotos_72350117-stock-illustration-no-user-profile-picture-hand.jpg',
  // },
]
