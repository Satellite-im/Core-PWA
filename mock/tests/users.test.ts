import * as Users from '../users'

describe('init', () => {
  it('gets the CallUsers constant', () => {
    expect(Users.CallUsers).toMatchSnapshot()
  })
  it('gets the User constant', () => {
    expect(Users.User).toMatchSnapshot()
  })
  it('gets the Users constant', () => {
    expect(Users.Users).toMatchSnapshot()
  })
})
