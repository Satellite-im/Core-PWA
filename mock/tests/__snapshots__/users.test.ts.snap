// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init gets the CallUsers constant 1`] = `
Array [
  Object {
    "address": "0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f",
    "last_message": "Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.",
    "name": "Taurus Nix",
    "state": "online",
    "status": "Some important status message",
  },
  Object {
    "address": "0x07ee55aa48bb72dcc6e9d78256648910de513eca",
    "last_message": "At the edge of forever bits of moving fluff radio telescope the carbon in our apple pies are creatures of the cosmos rings of Uranus.",
    "name": "Halley Themis",
    "state": "offline",
    "status": "Working on the space station",
  },
]
`;

exports[`init gets the User constant 1`] = `
Object {
  "address": "0x9bf4001d307dfd62b26a2f1307ee0c0307632d59",
  "last_message": "Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.",
  "name": "Lyra Cassini",
  "state": "online",
  "status": "Working on the space station",
}
`;

exports[`init gets the Users constant 1`] = `
Array [
  Object {
    "address": "0xc61b9bb3a7a0767e3179713f3a5c7a9aedce193c",
    "last_message": "Ship of the imagination consciousness across the centuries network of wormholes finite but unbounded inconspicuous motes of rock and gas.",
    "name": "Phoenix Kalindi",
    "profilePicture": "",
    "state": "online",
    "status": "Working on the space station",
    "unreads": 4,
  },
  Object {
    "address": "0xdc76cd25977e0a5ae17155770273ad58648900d3",
    "last_message": "Network of wormholes science citizens of distant epochs tingling of the spine cosmic ocean brain is the seed of intelligence. The sky calls to us two ghostly white figures in coveralls and helmets are softly dancing muse about paroxysm of global death the only home weve ever known hearts of the stars.",
    "name": "Ariel Larissa",
    "profilePicture": "",
    "state": "idle",
    "status": "Working on the space station",
    "unreads": 1,
  },
  Object {
    "address": "0xdf9eb223bafbe5c5271415c75aecd68c21fe3d7f",
    "badge": "verified",
    "last_message": "Apollonius of Perga laws of physics colonies paroxysm of global death Jean-François Champollion emerged into consciousness.",
    "name": "Taurus Nix",
    "state": "online",
    "status": "Some important status message",
  },
  Object {
    "address": "0x07ee55aa48bb72dcc6e9d78256648910de513eca",
    "last_message": "At the edge of forever bits of moving fluff radio telescope the carbon in our apple pies are creatures of the cosmos rings of Uranus.",
    "name": "Halley Themis",
    "state": "offline",
    "status": "Working on the space station",
  },
  Object {
    "address": "0x9bf4001d307dfd62b26a2f1307ee0c0307632d59",
    "last_message": "Tendrils of gossamer clouds stirred by starlight citizens of distant epochs network of wormholes corpus callosum the sky calls to us.",
    "name": "Lyra Cassini",
    "state": "idle",
    "status": "Working on the space station",
  },
  Object {
    "address": "0xdc76cd25977e0a5ae17155770273ad58648913d3",
    "last_message": "To be the first to enter the cosmos, to engage, single-handed, in an unprecedented duel with nature—could one dream of anything more?",
    "name": "Lady Calliope",
    "profilePicture": "",
    "state": "mobile",
    "status": "Working on the space station",
  },
  Object {
    "address": "0x07ee55aa48bb72dcc6e9d88856648910de513eca",
    "last_message": "At the edge of forever bits of moving fluff radio telescope the carbon in our apple pies are creatures of the cosmos rings of Uranus.",
    "name": "James Lightspeed",
    "state": "offline",
    "status": "Exploring deep space",
  },
  Object {
    "address": "0x9bf4001de07dfd62b26a2f1307ee0c0307632d59",
    "last_message": "Tendrils of gossamer clouds stirred by starlight citizens of distant epochs network of wormholes corpus callosum the sky calls to us.",
    "name": "Derek Wormhole",
    "state": "online",
    "status": "Filling space",
  },
  Object {
    "address": "0xdc76cd25977e015ae17155770273ad58648913d3",
    "last_message": "To be the first to enter the cosmos, to engage, single-handed, in an unprecedented duel with nature—could one dream of anything more?",
    "name": "Frank Meteor",
    "profilePicture": "",
    "state": "mobile",
    "status": "Working on the space station",
  },
]
`;
